connectors/sources/gmail.py

```diff
@@ -0,0 +1,352 @@
+        email = user.get(UserFields.EMAIL.value)
+        created_at = user.get(UserFields.CREATION_DATE.value)
+
+        return {
+            "_id": email,
+            "identity": {"email": email},
+            "created_at": created_at or iso_utc(),
+        } | self.access_control_query(access_control)
+
+    def _decorate_with_access_control(self, document, access_control):
+        if self._dls_enabled():
+            document[ACCESS_CONTROL] = list(
+                set(document.get(ACCESS_CONTROL, []) + access_control)
+            )
+
+        return document
+
+    async def get_access_control(self):
+        if not self._dls_enabled():
+            return
+
+        async for user in self._google_directory_client.users():
+            if user:
+                access_control = [user.get(UserFields.EMAIL.value)]
+
+                yield self._user_access_control_doc(user, access_control)
+
+    def _filtering_enabled(self, filtering):
+        return filtering is not None and filtering.has_advanced_rules()
+
+    async def _message_doc_with_access_control(
+        self, access_control, gmail_client, message
+    ):
+        message_id = message.get("id")
+        message_content = await gmail_client.message(message_id)
+        message_content["id"] = message_id
+
+        message_doc = _message_doc(message_content)
+        message_doc_with_access_control = self._decorate_with_access_control(
+            message_doc, access_control
+        )
+
+        return message_doc_with_access_control
+
+    async def get_docs(self, filtering=None):
+        if self._filtering_enabled(filtering):
+            self._logger.debug("Fetching documents using advanced rules.")
+
+            advanced_rules = filtering.get_advanced_rules()
+            message_queries = advanced_rules.get("messages", [])
+
+            async for user in self._google_directory_client.users():
+                email = user.get(UserFields.EMAIL.value)
+                access_control = [email]
+
+                # reinitialization is needed to work around a 403 Forbidden error (see: https://issuetracker.google.com/issues/290567932)
+                gmail_client = self._gmail_client(email)
+
+                for message_query in message_queries:
+                    self._logger.debug(f"Fetching messages for query: {message_query}.")
+
+                    async for message in gmail_client.messages(query=message_query):
+                        if not message:
+                            continue
+
+                        message_doc_with_access_control = (
+                            await self._message_doc_with_access_control(
+                                access_control, gmail_client, message
+                            )
+                        )
+
+                        yield message_doc_with_access_control, None
+        else:
+            async for user in self._google_directory_client.users():
```

Comment:
If user A sends an email to the whole organization (assuming 2000 users), does it mean the same message will be indexed for 2000 times?